### DISCUSSION TITLE
defect analysis as an add-on package

### DISCUSSION THREAD

**Discussion by jmmshn** [materialsproject/pymatgen#2383] [Ideas]
Hi @mkhorton @shyuep,
I'm working on adding some defect analysis code that will have a few new dependencies like (`dscribe` and `mp-pyrho`)
I'm wondering if it makes sense to separate the `analysis.defects` as an add-on package since I heard we are worried about additional dependencies.

As far as I can tell the defect code is used only in one other places `strucutre_matcher.py` so it should not be so disruptive, also I think a few of the charge density-based interstitial tests are pretty time-consuming so that is a possible plus as well

**Comment by mkhorton** [comment]
Also `sk-image` as pointed out by @acrutt

**Comment by shyuep** [comment]
I definitely support moving defect analysis to a add-on package. As far as I know, the only place where the structure matcher uses it is in the DefectComparator. That can definitely be either moved back to analysis.defects or alternatively, can be surrounded by a try-catch to inform users to install the pymatgen-analysis-defects package if they want to use it.

**Comment by shyuep** [comment]
FYI, pls refer to https://github.com/materialsproject/pymatgen-addon-template for how to make the addon.

**Comment by mkhorton** [comment]
I already pointed this out to @jmmshn in separate conversation, but there is an issue with `numba` (a dep of `dscribe` and others) being fundamentally incompatible with the latest pymatgen due to not supporting the latest `numpy`.

It's tricky, because `numba` is used by a lot of packages (eg, I know the developers of `py4dstem` want to use `pymatgen` but also want to use `numba` and it's causing them issues).

**Comment by mkhorton** [comment]
In the context of this conversation, having defects be a separate package will at least mean these version incompatibly issues aren't affecting pymatgen core, but they will still have to be dealt with.

**Comment by shyuep** [comment]
I would recommend that everyone who needs to work with packages requiring old numpy to just use pymatgen 2022.0.x (specifically 2022.0.17). As far as I know, all those still work. I am open to ideas as to how to solve the numpy issue. I have no idea why binary incompatibilities exist between different versions of numpy. It has already happened a few times in the past one year.
